key.go

```diff
@@ -0,0 +1,778 @@
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"errors"
+	"fmt"
+	"math/big"
+	"strconv"
+
+	cbor "github.com/fxamacker/cbor/v2"
+)
+
+const (
+	// The key is used to create signatures. Requires private key fields.
+	KeyOpSign KeyOp = 1
+
+	// The key is used for verification of signatures.
+	KeyOpVerify KeyOp = 2
+
+	// the key is used for key transport encryption.
+	KeyOpEncrypt KeyOp = 3
+
+	// the key is used for key transport decryption. Requires private key fields.
+	KeyOpDecrypt KeyOp = 4
+
+	// The key is used for key wrap encryption.
+	KeyOpWrapKey KeyOp = 5
+
+	// The key is used for key wrap decryption.
+	KeyOpUnwrapKey KeyOp = 6
+
+	// The key is used for deriving keys. Requires private key fields.
+	KeyOpDeriveKey KeyOp = 7
+
+	// The key is used for deriving bits not to be used as a key. Requires
+	// private key fields.
+	KeyOpDeriveBits KeyOp = 8
+
+	// The key is used for creating MACs.
+	KeyOpMacCreate KeyOp = 9
+
+	// The key is used for validating MACs.
+	KeyOpMacVerify KeyOp = 10
+)
+
+// KeyOp represents a key_ops value used to restrict purposes for which a Key
+// may be used.
+type KeyOp int64
+
+// KeyOpFromString returns the KeyOp corresponding to the specified name.
+func KeyOpFromString(val string) (KeyOp, error) {
+	switch val {
+	case "sign":
+		return KeyOpSign, nil
+	case "verify":
+		return KeyOpVerify, nil
+	case "encrypt":
+		return KeyOpEncrypt, nil
+	case "decrypt":
+		return KeyOpDecrypt, nil
+	case "wrap key":
+		return KeyOpWrapKey, nil
+	case "unwrap key":
+		return KeyOpUnwrapKey, nil
+	case "derive key":
+		return KeyOpDeriveKey, nil
+	case "derive bits":
+		return KeyOpDeriveBits, nil
+	case "MAC create":
+		return KeyOpMacCreate, nil
+	case "MAC verify":
+		return KeyOpMacVerify, nil
+	default:
+		return -1, fmt.Errorf("unknown key_ops value %q", val)
+	}
+}
+
+func (ko KeyOp) String() string {
+	switch ko {
+	case KeyOpSign:
+		return "sign"
+	case KeyOpVerify:
+		return "verify"
+	case KeyOpEncrypt:
+		return "encrypt"
+	case KeyOpDecrypt:
+		return "decrypt"
+	case KeyOpWrapKey:
+		return "wrap key"
+	case KeyOpUnwrapKey:
+		return "unwrap key"
+	case KeyOpDeriveKey:
+		return "derive key"
+	case KeyOpDeriveBits:
+		return "derive bits"
+	case KeyOpMacCreate:
+		return "MAC create"
+	case KeyOpMacVerify:
+		return "MAC verify"
+	default:
+		return "unknown key_op value " + strconv.Itoa(int(ko))
+
+	}
+}
+
+// MarshalCBOR marshals the KeyOp as a CBOR int.
+func (ko KeyOp) MarshalCBOR() ([]byte, error) {
+	return encMode.Marshal(int(ko))
+}
+
+// UnmarshalCBOR populates the KeyOp form the provided CBOR value (must be int
+// or tstr).
+func (ko *KeyOp) UnmarshalCBOR(data []byte) error {
+	var raw IntOrStr
+
+	if err := raw.UnmarshalCBOR(data); err != nil {
+		return fmt.Errorf("invalid key_ops value %w", err)
+	}
+
+	if raw.IsString() {
+		v, err := KeyOpFromString(raw.String())
+
+		if err != nil {
+			return err
+		}
+
+		*ko = v
+	} else {
+		v := raw.Int()
+
+		if v < 1 || v > 10 {
```

Comment:
similar to the comment I made for the `Algorithm` counterpart (i.e., tighter integration with the type)